File: jambeth/jAmbeth.Xml/src/main/java/de/osthus/ambeth/xml/simple/SimpleXmlController.java
package de.osthus.ambeth.xml.simple;

import de.osthus.ambeth.ioc.annotation.Autowired;
import de.osthus.ambeth.ioc.extendable.ClassExtendableContainer;
import de.osthus.ambeth.ioc.extendable.MapExtendableContainer;
import de.osthus.ambeth.merge.IProxyHelper;
import de.osthus.ambeth.xml.ICyclicXmlController;
import de.osthus.ambeth.xml.INameBasedHandler;
import de.osthus.ambeth.xml.INameBasedHandlerExtendable;
import de.osthus.ambeth.xml.IReader;
import de.osthus.ambeth.xml.ITypeBasedHandler;
import de.osthus.ambeth.xml.ITypeBasedHandlerExtendable;
import de.osthus.ambeth.xml.IWriter;

public class SimpleXmlController implements ICyclicXmlController, ITypeBasedHandlerExtendable, INameBasedHandlerExtendable
{
	protected final ClassExtendableContainer<ITypeBasedHandler> typeToElementHandlers = new ClassExtendableContainer<ITypeBasedHandler>("elementHandler",
			"type");

	protected final MapExtendableContainer<String, INameBasedHandler> nameToElementHandlers = new MapExtendableContainer<String, INameBasedHandler>(
			"elementHandler", "name");

	@Autowired
	protected IProxyHelper proxyHelper;

	@Override
	public Object readObject(IReader reader)
	{
		return readObject(Object.class, reader);
	}

	@Override
	public Object readObject(Class<?> returnType, IReader reader)
	{
		String elementName = reader.getElementName();
		INameBasedHandler nameBasedHandler = nameToElementHandlers.getExtension(elementName);
		if (nameBasedHandler == null)
		{
			throw new IllegalArgumentException("Could not read object: " + elementName);
		}
		return nameBasedHandler.readObject(returnType, elementName, 0, reader);
	}

	@Override
	public void writeObject(Object obj, IWriter writer)
	{
		Class<?> type = proxyHelper.getRealType(obj.getClass());
		ITypeBasedHandler extension = typeToElementHandlers.getExtension(type);
		if (extension == null)
		{
			throw new IllegalArgumentException("Could not write object: " + obj);
		}
		extension.writeObject(obj, type, writer);
	}

	@Override
	public void registerElementHandler(ITypeBasedHandler elementHandler, Class<?> type)
	{
		typeToElementHandlers.register(elementHandler, type);
	}

	@Override
	public void unregisterElementHandler(ITypeBasedHandler elementHandler, Class<?> type)
	{
		typeToElementHandlers.unregister(elementHandler, type);
	}

	@Override
	public void registerNameBasedElementHandler(INameBasedHandler nameBasedElementHandler, String elementName)
	{
		nameToElementHandlers.register(nameBasedElementHandler, elementName);
	}

	@Override
	public void unregisterNameBasedElementHandler(INameBasedHandler nameBasedElementHandler, String elementName)
	{
		nameToElementHandlers.unregister(nameBasedElementHandler, elementName);
	}
}

File: ambeth/Ambeth.Xml/ambeth/xml/simple/SimpleXmlController.cs
using De.Osthus.Ambeth.Ioc.Annotation;
using De.Osthus.Ambeth.Ioc.Extendable;
using De.Osthus.Ambeth.Merge;
using System;
namespace De.Osthus.Ambeth.Xml.Simple
{
    public class SimpleXmlController : ICyclicXmlController, ITypeBasedHandlerExtendable, INameBasedHandlerExtendable
    {
        protected readonly ClassExtendableContainer<ITypeBasedHandler> typeToElementHandlers = new ClassExtendableContainer<ITypeBasedHandler>("elementHandler",
                "type");

        protected readonly MapExtendableContainer<String, INameBasedHandler> nameToElementHandlers = new MapExtendableContainer<String, INameBasedHandler>(
                "elementHandler", "name");

        [Autowired]
        public IProxyHelper ProxyHelper { protected get; set; }

        public Object ReadObject(IReader reader)
        {
            return ReadObject(typeof(Object), reader);
        }

        public Object ReadObject(Type returnType, IReader reader)
        {
            String elementName = reader.GetElementName();
            INameBasedHandler nameBasedHandler = nameToElementHandlers.GetExtension(elementName);
            if (nameBasedHandler == null)
            {
                throw new ArgumentException("Could not read object: " + elementName);
            }
            return nameBasedHandler.ReadObject(returnType, elementName, 0, reader);
        }

        public void WriteObject(Object obj, IWriter writer)
        {
            Type type = ProxyHelper.GetRealType(obj.GetType());
            ITypeBasedHandler extension = typeToElementHandlers.GetExtension(type);
            if (extension == null)
            {
                throw new ArgumentException("Could not write object: " + obj);
            }
            extension.WriteObject(obj, type, writer);
        }

        public void RegisterElementHandler(ITypeBasedHandler elementHandler, Type type)
        {
            typeToElementHandlers.Register(elementHandler, type);
        }

        public void UnregisterElementHandler(ITypeBasedHandler elementHandler, Type type)
        {
            typeToElementHandlers.Unregister(elementHandler, type);
        }

        public void RegisterNameBasedElementHandler(INameBasedHandler nameBasedElementHandler, String elementName)
        {
            nameToElementHandlers.Register(nameBasedElementHandler, elementName);
        }

        public void UnregisterNameBasedElementHandler(INameBasedHandler nameBasedElementHandler, String elementName)
        {
            nameToElementHandlers.Unregister(nameBasedElementHandler, elementName);
        }
    }
}